observer/lib/observer.js

```diff
@@ -0,0 +1,39 @@
+import { updateDailyTransferStats } from './platform-stats.js'
+
+/**
+ * Observe the transfer events on the Filecoin blockchain
+ * @param {import('pg').Pool} pgPool
+ * @param {import('ethers').Contract} ieContract
+ * @param {import('ethers').Provider} provider
+ */
+export const observeTransferEvents = async (pgPool, ieContract, provider) => {
+  const { rows } = await pgPool.query(
+    'SELECT MAX(last_checked_block) FROM daily_reward_transfers'
+  )
+  const lastCheckedBlock = rows[0].last_checked_block
+
+  console.log('Querying impact evaluator Transfer events after block', lastCheckedBlock)
+  let events
+  try {
+    events = await ieContract.queryFilter(ieContract.filters.Transfer(), lastCheckedBlock)
+  } catch (error) {
+    console.error('Error querying impact evaluator Transfer events', error)
+    if (error.message.includes('bad tipset height')) {
+      console.log('Block number too old, GLIF only provides last 2000 blocks, querying from -1900')
+      events = await ieContract.queryFilter(ieContract.filters.Transfer(), -1900)
+    } else {
+      throw error
+    }
+  }
+  const currentBlockNumber = await provider.getBlockNumber()
```

Comment:
Since we are querying the RPC API for the last block number, can we change the queryFilter range logic to be smarter and use the last block number to decide how many blocks we want to query?

Something along the following lines:

```js
const currentBlockNumber = await provider.getBlockNumber()
let queryFrom = lastCheckedBlock
if (queryFrom > currentBlockNumber - 1900) {
  console.log('Block number too old, GLIF only provides last 2000 blocks, querying from -1900')
  queryFrom = currentBlockNumber - 1900
}
```

Then we don't need the fragile check detecting "bad tipset height" error.

Just an idea to consider.

Reply:
Great idea! I also didn't love the error message text check, but the response didn't have any more concrete error code unfortunately.

But I should have thought of this, makes much more sense :)

Reply:
Addressed in #134, along with below comments (except batching, awaiting your response about that)